DirectProgramming/C++SYCL_FPGA/Tutorials/Features/ip_authoring_interfaces/component_interfaces_comparison/csr-pipes/src/vector_add.cpp

```diff
@@ -0,0 +1,121 @@
+#include <iostream>
+
+// oneAPI headers
+#include <sycl/sycl.hpp>
+#include <sycl/ext/intel/fpga_extensions.hpp>
+#include "exception_handler.hpp"
+
+// Forward declare the kernel name in the global scope. This is an FPGA best
+// practice that reduces name mangling in the optimization reports.
+class SimpleVAddPipes;
+
+// Forward declare pipe names to reduce name mangling (maybe get rid of; pipes
+// already have bad names)
+class IDPipeA;
+class IDPipeB;
+class IDPipeC;
+
+#define VECT_SIZE 256
+
+using PipeProps = decltype(sycl::ext::oneapi::experimental::properties(
+    sycl::ext::intel::experimental::ready_latency<0>));
+
+using InputPipeA =
+    sycl::ext::intel::experimental::pipe<IDPipeA, int, 0,
+                                         PipeProps>;
+using InputPipeB =
+    sycl::ext::intel::experimental::pipe<IDPipeB, int, 0,
+                                         PipeProps>;
+
+using CSRPipeProps = decltype(sycl::ext::oneapi::experimental::properties(
+    sycl::ext::intel::experimental::protocol_avalon_mm));
+
+// this csr pipe will only be read and write once with the total sum
+using OutputPipeC =
+    sycl::ext::intel::experimental::pipe<IDPipeC, int, 0,
+                                         CSRPipeProps>;
+
+struct SimpleVAddKernelPipes {
+  int len;
+
+  void operator()() const {
+    int sum_total = 0;
+    for (int idx = 0; idx < len; idx++) {
+      int a_val = InputPipeA::read();
+      int b_val = InputPipeB::read();
+      int sum = a_val + b_val;
+
+      sum_total += sum;
+    }
+
+    // Write to OutputPipeC only once per kernel invocation
+    OutputPipeC::write(sum_total);
+  }
+};
+
+int main() {
+  try {
+    // Use compile-time macros to select either:
+    //  - the FPGA emulator device (CPU emulation of the FPGA)
+    //  - the FPGA device (a real FPGA)
+    //  - the simulator device
+#if FPGA_SIMULATOR
+    auto selector = sycl::ext::intel::fpga_simulator_selector_v;
+#elif FPGA_HARDWARE
+    auto selector = sycl::ext::intel::fpga_selector_v;
+#else  // #if FPGA_EMULATOR
+    auto selector = sycl::ext::intel::fpga_emulator_selector_v;
+#endif
+
+    // create the device queue
+    sycl::queue q(selector, fpga_tools::exception_handler);
+
+    int count = VECT_SIZE;  // pass array size by value
+
+    int expected_sum = 0;
+
+    // push data into pipes
+    int *a = new int[count];
+    int *b = new int[count];
+    for (int i = 0; i < count; i++) {
+      a[i] = i;
+      b[i] = (count - i);
+
+      expected_sum += (a[i] + b[i]);
+
+      InputPipeA::write(q, a[i]);
+      InputPipeB::write(q, b[i]);
+    }
+
+    std::cout << "Add two vectors of size " << count << std::endl;
+
+    q.single_task<SimpleVAddPipes>(SimpleVAddKernelPipes{count});
```

Comment:
You might need a wait() call after the kernel to make sure it finishes. Let's discuss this more before making the change...not sure if it works correctly without it (it might, if OutputPipeC::read() will block)

```suggestion
    q.single_task<SimpleVAddPipes>(SimpleVAddKernelPipes{count}).wait();
```

Reply:
would the host being blocked on the blocking read until kernel finishes anyway?